File: android/app/src/main/java/my/ssdid/wallet/domain/profile/ProfileManager.kt
package my.ssdid.wallet.domain.profile

import my.ssdid.wallet.domain.model.CredentialSubject
import my.ssdid.wallet.domain.model.Proof
import my.ssdid.wallet.domain.model.VerifiableCredential
import my.ssdid.wallet.domain.vault.Vault
import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import javax.inject.Inject

class ProfileManager @Inject constructor(private val vault: Vault) {

    companion object {
        const val PROFILE_ID = "urn:ssdid:profile"
        const val SELF_ISSUER = "did:ssdid:self"
    }

    suspend fun saveProfile(name: String, email: String, phone: String): Result<Unit> = runCatching {
        // Delete existing profile if present
        val existing = vault.listCredentials().find { it.id == PROFILE_ID }
        if (existing != null) vault.deleteCredential(PROFILE_ID).getOrThrow()

        val now = DateTimeFormatter.ISO_INSTANT.format(Instant.now().atOffset(ZoneOffset.UTC))
        val claims = mutableMapOf("name" to name, "email" to email)
        if (phone.isNotBlank()) claims["phone"] = phone

        val credential = VerifiableCredential(
            id = PROFILE_ID,
            type = listOf("VerifiableCredential", "ProfileCredential"),
            issuer = SELF_ISSUER,
            issuanceDate = now,
            credentialSubject = CredentialSubject(id = SELF_ISSUER, claims = claims),
            proof = Proof(
                type = "SelfIssued2024",
                created = now,
                verificationMethod = SELF_ISSUER,
                proofPurpose = "selfAssertion",
                proofValue = ""
            )
        )
        vault.storeCredential(credential).getOrThrow()
    }

    suspend fun getProfile(): VerifiableCredential? {
        return vault.listCredentials().find { it.id == PROFILE_ID }
    }

    suspend fun getProfileClaims(): Map<String, String> {
        return getProfile()?.credentialSubject?.claims ?: emptyMap()
    }
}

File: ios/SsdidWallet/Domain/Profile/ProfileManager.swift
import Foundation

/// Manages user profile data stored as a self-issued verifiable credential.
final class ProfileManager {

    static let profileId = "urn:ssdid:profile"
    static let selfIssuer = "did:ssdid:self"

    private let vault: Vault

    init(vault: Vault) {
        self.vault = vault
    }

    /// Saves or updates the user's profile (name, email) as a self-issued VC.
    func saveProfile(name: String, email: String) async throws {
        // Delete existing profile if present
        let existing = await vault.listCredentials()
        if existing.contains(where: { $0.id == Self.profileId }) {
            try await vault.deleteCredential(credentialId: Self.profileId)
        }

        let now = ISO8601DateFormatter().string(from: Date())
        let claims = ["name": name, "email": email]

        let credential = VerifiableCredential(
            id: Self.profileId,
            type: ["VerifiableCredential", "ProfileCredential"],
            issuer: Self.selfIssuer,
            issuanceDate: now,
            credentialSubject: CredentialSubject(id: Self.selfIssuer, claims: claims),
            proof: Proof(
                type: "SelfIssued2024",
                created: now,
                verificationMethod: Self.selfIssuer,
                proofPurpose: "selfAssertion",
                proofValue: ""
            )
        )

        try await vault.storeCredential(credential)
    }

    /// Retrieves the user's profile credential.
    func getProfile() async -> VerifiableCredential? {
        let credentials = await vault.listCredentials()
        return credentials.first { $0.id == Self.profileId }
    }

    /// Retrieves the user's profile claims (name, email, etc.).
    func getProfileClaims() async -> [String: String] {
        guard let profile = await getProfile() else { return [:] }
        return profile.credentialSubject.claims
    }
}
